mesonbuild/modules/java.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pathlib
+from mesonbuild.mesonlib.universal import File
+from mesonbuild.build import CustomTarget
+from mesonbuild.optinterpreter import permitted_kwargs
+import os
+from mesonbuild.interpreterbase.decorators import FeatureNew, typed_pos_args
+import typing as T
+from . import ExtensionModule, ModuleReturnValue, ModuleState
+from ..interpreter import Interpreter
+from mesonbuild.interpreter.interpreterobjects import FileHolder
+
+class JavaModule(ExtensionModule):
+    @FeatureNew('Java Module', '0.59.0')
+    def __init__(self, interpreter: Interpreter):
+        super().__init__(interpreter)
+        self.methods.update({
+            'generate_native_header': self.generate_native_header,
+        })
+
+    @typed_pos_args('generate_native_header', (str, FileHolder), min_varargs=1, max_varargs=1)
+    # @permitted_kwargs({'package'})
```

Comment:
I do not understand why this doesn't work. Help is appreciated. When uncommented...

```
Traceback (most recent call last):
  File "/home/tristan957/Projects/meson/mesonbuild/mesonmain.py", line 223, in run
    return options.run_func(options)
  File "/home/tristan957/Projects/meson/mesonbuild/msetup.py", line 281, in run
    app.generate()
  File "/home/tristan957/Projects/meson/mesonbuild/msetup.py", line 184, in generate
    self._generate(env)
  File "/home/tristan957/Projects/meson/mesonbuild/msetup.py", line 223, in _generate
    intr.run()
  File "/home/tristan957/Projects/meson/mesonbuild/interpreter/interpreter.py", line 2424, in run
    super().run()
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 132, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 157, in evaluate_codeblock
    raise e
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 150, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 163, in evaluate_statement
    return self.function_call(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 549, in function_call
    res = func(node, func_args, kwargs)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/decorators.py", line 598, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/decorators.py", line 102, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreter/interpreter.py", line 1909, in func_subdir
    self.evaluate_codeblock(codeblock)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 157, in evaluate_codeblock
    raise e
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 150, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 165, in evaluate_statement
    self.assignment(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 910, in assignment
    value = self.evaluate_statement(node.value)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 167, in evaluate_statement
    return self.method_call(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 593, in method_call
    return self._holderify(obj.method_call(method_name, args, kwargs))
  File "/home/tristan957/Projects/meson/mesonbuild/interpreter/interpreterobjects.py", line 769, in method_call
    ret = method(state, args, kwargs)
TypeError: _inner() takes 3 positional arguments but 4 were given
```

Reply:
I don't understand how 4 arguments were passed though.

Reply:
You don't want `permitted_kwargs` anway, that's old.
```python
@typed_kwargs(
  'java.generate_native_header',
  KwargInfo('package', (str, None)),
)
def ...:
   package = kwargs['package']
```

Reply:
Sorry, I just noticed that you were using permitted_kwargs...